doc/source/serve/deployment-graph.md

```diff
@@ -0,0 +1,545 @@
+
+## Key Features
+- Provide the ability to author a DAG of Serve nodes to form a complex inference graph.
+- Graph authoring experience should be fully Python-programmable with support for dynamic selection, control flows, user business logic, etc.
+- Graph can be instantiated and locally executed using tasks and actors API
+- Graph can be deployed as a group where individual nodes can be reconfigured and scaled indepenently.
+
+__[Full Ray Enhancement Proposal, REP-001: Serve Pipeline](https://github.com/ray-project/enhancements/blob/main/reps/2022-03-08-serve_pipeline.md)__
+
++++
+
+## Concepts
+
+### Deployment
+Scalable, upgradeable group of actors managed by Ray Serve. __[See docs for detail](https://docs.ray.io/en/master/serve/core-apis.html#core-api-deployments)__
+
+### DeploymentNode
+Smallest unit in a graph, typically a serve annotated class or function, backed by a Deployment.
+
+### InputNode 
+Singleton node in a graph that holds user input value, which is resolved at runtime of graph execution.
+
+### Bind
+A graph building API applicable to decorated class or function.  `decorated_class_or_func.bind(*args, **kwargs)` generates an IR node that can be used to build graph, and bound arguments will be applied at execution time, including dynamic user input.
+
+### Deployment Graph
+Collection of deployment nodes bound together that forms a DAG that represents an inference graph for complicated tasks, can be deployed and call as a unit. For example: ensemble, chaining, dynamic selection. 
+
+
++++
+
+## Full End to End Example Walkthrough
+
+Let's put the concepts together and incrementally build a runnable DAG example including the following attributes:
+
+```{tip}
+At the end of this document we have the full and end to end executable implementation.
+```
+
+- All nodes in the deployment graph naturally forms a DAG structure.
+- A node could use or call into other nodes in the deployment graph.
+- Deployment graph has mix of class and function as nodes.
+- Same input or output can be used in multiple nodes in the DAG.
+- A node might access partial user input.
+- Control flow is used where dynamic dispatch happens with respect to input value.
+- Same class can be constructed, or function bound with different args that generates multiple distinct nodes in DAG.
+- A node can be called either sync or async.
+
++++
+
+![deployment graph](https://github.com/ray-project/images/blob/master/docs/serve/deployment_graph.png?raw=true)
+
++++
+
+### Step 1: User InputNode and preprocessor 
+
+Let's start with the first layer of DAG: Building user input to two preprocessor functions, where each function receives parts of the input data. For simplicity, we use the same existing `@serve.deployment` decorator on an async function body.
+
++++
+
+```python
+from ray import serve
+from ray.experimental.dag.input_node import InputNode
+
+@serve.deployment
+async def preprocessor(input_data: str):
```

Comment:
i want to show an async example with blocking computations that user might not expect, then either do `asyncio.gather`  or showing user can directly bind an async function. I think alternatively i can do `await asyncio.sleep()` to yield control but im more in favor of the current one since user might not put heavy computation with right `await` syntax in actual implementation.

Reply:
+1 for await asyncio. you can show the async def somewhere else (in particular the driver node)

Reply:
updated